File: crates/facet-generate/src/generation/typescript/emitter/features/MapOfT.ts
function serializeMap<K, V>(
    value: Map<K, V>,
    serializer: Serializer,
    serializeEntry: (key: K, value: V, serializer: Serializer) => void,
): void {
    serializer.serializeLen(value.size);
    const offsets: number[] = [];
    for (const [k, v] of value.entries()) {
        offsets.push(serializer.getBufferOffset());
        serializeEntry(k, v, serializer);
    }
    serializer.sortMapEntries(offsets);
}

function deserializeMap<K, V>(
    deserializer: Deserializer,
    deserializeEntry: (deserializer: Deserializer) => [K, V],
): Map<K, V> {
    const length = deserializer.deserializeLen();
    const obj = new Map<K, V>();
    for (let i = 0; i < length; i++) {
        const [key, value] = deserializeEntry(deserializer);
        obj.set(key, value);
    }
    return obj;
}

File: crates/facet_generate/src/generation/swift/emitter/features/MapOfT.swift
func serializeMap<K, V, S: Serializer>(
    value: [K: V],
    serializer: S,
    serializeEntry: (K, V, S) throws -> Void
) throws {
    try serializer.serialize_len(value: value.count)
    for (key, value) in value {
        try serializeEntry(key, value, serializer)
    }
}

func deserializeMap<K: Hashable, V, D: Deserializer>(
    deserializer: D,
    deserializeEntry: (D) throws -> (K, V)
) throws -> [K: V] {
    let length = try deserializer.deserialize_len()
    var obj: [K: V] = [:]
    for _ in 0..<length {
        let (key, value) = try deserializeEntry(deserializer)
        obj[key] = value
    }
    return obj
}
